src/utils/kamu-cli-puppet-ext/src/kamu_cli_puppet_ext.rs

```diff
@@ -0,0 +1,211 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::io::Write;
+use std::net::Ipv4Addr;
+use std::path::PathBuf;
+
+use async_trait::async_trait;
+use datafusion::prelude::{ParquetReadOptions, SessionContext};
+use kamu_cli_puppet::KamuCliPuppet;
+use opendatafabric::serde::yaml::YamlDatasetSnapshotSerializer;
+use opendatafabric::serde::DatasetSnapshotSerializer;
+use opendatafabric::{DatasetName, DatasetRef, DatasetSnapshot};
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[async_trait]
+pub trait KamuCliPuppetExt {
+    async fn get_dataset_names(&self) -> Vec<String>;
+
+    async fn add_dataset(&self, dataset_snapshot: DatasetSnapshot);
+
+    async fn get_list_of_repo_aliases(
+        &self,
+        dataset_ref: &DatasetRef,
+    ) -> (Vec<String>, Vec<String>);
+
+    async fn complete<T>(&self, input: T, current: usize) -> Vec<String>
+    where
+        T: Into<String> + Send;
+
+    async fn start_api_server(self, e2e_data_file_path: PathBuf) -> ServerOutput;
+
+    async fn assert_last_data_slice(
+        &self,
+        dataset_name: &DatasetName,
+        expected_schema: &str,
+        expected_data: &str,
+    );
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[async_trait]
+impl KamuCliPuppetExt for KamuCliPuppet {
+    async fn get_dataset_names(&self) -> Vec<String> {
+        let assert = self
+            .execute(["list", "--output-format", "csv"])
+            .await
+            .success();
+
+        let stdout = std::str::from_utf8(&assert.get_output().stdout).unwrap();
+
+        stdout
+            .lines()
+            .skip(1) // Skip header
+            .map(|line| line.split(',').next().unwrap().to_string())
```

Comment:
I would also try to keep methods as close to CLI command names as possible, e.g. instead of `get_dataset_names` have `list_datasets` and let caller filter the data they need from the structs. We don't want to create a new semantically different API.

Reply:
> Why parse CSV manually?

looks like some flashbacks from my Advent of Code participations

--

In all seriousness, yes, that makes sense -- fixed